### PR TITLE
timeline for distributed training

### PR DESCRIPTION
**Pull request by panyx0718**
One trainer task is enable for profiling. The trainer will tell PS servers to start profiling. When the trainer stops profiling, it also informs the PS servers. They all generate profile files that can be converted to timeline

image

### CODE REVIEW

File: cmake/external/grpc.cmake [hunk 1]
```diff
@@ -33,7 +33,7 @@ ExternalProject_Add(
     extern_grpc
     DEPENDS protobuf zlib
     GIT_REPOSITORY "https://github.com/grpc/grpc.git"
-    GIT_TAG "v1.10.x"
+    GIT_TAG "v1.8.x"
```
Comment (typhoonzero) [inline cmake/external/grpc.cmake]: I'm still looking into this issue https://github.com/PaddlePaddle/Paddle/issues/10153, version `1.10.x` should have some performance boost compared to `1.8.x`, if the machine you are using encounters this issue, can you please use another one for test?

File: paddle/fluid/operators/detail/send_recv.proto [hunk 2]
```diff
@@ -69,6 +69,10 @@ message VariableMessage {
   bytes rows = 9;
   // Look up table block execution output variable name.
   string out_varname = 10;
+  // If true, the ps server will start profiling, the ps
+  // server stops profiling and generates a profile to /tmp/profile_ps_*
+  // when profile switches from true to false.
+  bool profile = 11;
```
Comment (typhoonzero) [inline paddle/fluid/operators/detail/send_recv.proto]: Can we just use a envronment variable to determine whether to run profile?

File: cmake/external/grpc.cmake (hunk 1 above)
Comment (panyx0718) [reply cmake/external/grpc.cmake]: Done

File: paddle/fluid/operators/detail/send_recv.proto (hunk 2 above)
Comment (panyx0718) [reply paddle/fluid/operators/detail/send_recv.proto]: Using env is difficult to start and stop profiling of multiple machines at the same time.

File: benchmark/cluster/vgg16/vgg16_fluid.py [hunk 3]
```diff
@@ -80,6 +80,8 @@ def str2bool(v):
     type=str,
     default="",
     help="Comma-separated list of hostname:port pairs")
+parser.add_argument(
+    "--profile", action='store_true', help="If set, profile a few steps.")
```
Comment (jacquesqiao) [inline benchmark/cluster/vgg16/vgg16_fluid.py]: what is the meaning of `profile a few steps`?

File: paddle/fluid/operators/detail/sendrecvop_utils.cc [hunk 4]
```diff
@@ -45,6 +46,13 @@ void SerializeToByteBuffer(const std::string& name, framework::Variable* var,
   void* payload = nullptr;
   size_t payload_size;
   ProtoEncodeHelper e(static_cast<char*>(buf), 1024);
+  // Note: normally the profiler is enabled in 1 trainer, hence only
```
Comment (jacquesqiao) [inline paddle/fluid/operators/detail/sendrecvop_utils.cc]: How to ensure that only one trainer is enabled?

File: paddle/fluid/operators/listen_and_serv_op.cc [hunk 5]
```diff
@@ -294,6 +295,8 @@ void ListenAndServOp::RunAsyncLoop(framework::Executor *executor,
 
 void ListenAndServOp::RunImpl(const framework::Scope &scope,
                               const platform::Place &dev_place) const {
+  // Mark this as PS that it should decide profiling by listening from trainer.
+  platform::SetProfileLisener();
```
[the remark this answers is not in the document]
Comment (panyx0718) [reply paddle/fluid/operators/listen_and_serv_op.cc]: Done

File: paddle/fluid/operators/detail/sendrecvop_utils.cc (hunk 4 above)
Comment (panyx0718) [reply paddle/fluid/operators/detail/sendrecvop_utils.cc]: see vgg_fluid example. we can use args.task_index == 0 to verify this.

File: benchmark/cluster/vgg16/vgg16_fluid.py [hunk 6]
```diff
@@ -196,14 +198,28 @@ def train_loop(exe, trainer_prog):
                     feed={"pixel": img_data,
                           "label": y_data},
                     fetch_list=[avg_cost, batch_acc, batch_size])
+                return loss, acc, b_size
+
+            if args.profile and args.task_index == 0:
```
Comment (jacquesqiao) [inline benchmark/cluster/vgg16/vgg16_fluid.py]: maybe here we can use trainer_id == 0

Comment (panyx0718) [reply benchmark/cluster/vgg16/vgg16_fluid.py]: I think task_index and trainer_id mean the same thing here? vgg_fluid.py only has task_index, but no trainer_id.

File: paddle/fluid/operators/detail/variable_response.cc [hunk 7]
```diff
@@ -427,7 +428,26 @@ int VariableResponse::Parse(Source* source) {
...
         break;
       }
-
+      case sendrecv::VariableMessage::kProfileFieldNumber: {
+        bool profiling;
+        if (!input.ReadRaw(reinterpret_cast<void*>(&profiling), 1)) {
+          return tag;
+        }
+        meta_.set_profile(profiling);
+        int64_t lisner_id = platform::ListenerId();
```
Comment (jacquesqiao) [inline paddle/fluid/operators/detail/variable_response.cc]: lisner_id  => listener_id

Comment (panyx0718) [reply paddle/fluid/operators/detail/variable_response.cc]: Done

**Pull request merged**